httpcore/_async/http_proxy.py

```diff
@@ -184,3 +204,36 @@ def __init__(
             response[4], connection=connection, callback=self._response_closed
         )
         return response[0], response[1], response[2], response[3], wrapped_stream
+
+    async def _socks4_request(
+        self,
+        method: bytes,
+        url: URL,
+        headers: Headers = None,
+        stream: AsyncByteStream = None,
+        timeout: TimeoutDict = None,
+    ) -> Tuple[bytes, int, bytes, Headers, AsyncByteStream]:
+        """
+        SOCKS4 requires negotiation with the proxy.
+        """
+        origin = url[:3]
+        connection = await self._get_connection_from_pool(origin)
+
+        if connection is None:
+            connection = AsyncSOCKSConnection(origin, self.proxy_origin, "SOCKS4")
+            async with self._thread_lock:
+                self._connections.setdefault(origin, set())
+                self._connections[origin].add(connection)
+
+        # Issue a forwarded proxy request...
+
+        # GET https://www.example.org/path HTTP/1.1
```

Comment:
Actually this does not work with HTTPS. Changing the port to 443 and the scheme to `https` in the example fails with: `httpcore._exceptions.ProtocolError: can't handle event type ConnectionClosed when role=SERVER and state=SEND_RESPONSE`.

<details>
Sending  Request(method=b'GET', target=b'/', headers=[(b'host', b'example.org')], http_version=b'1.1')
Sending  Data(data=b'', chunk_start=False, chunk_end=False)
Sending  EndOfMessage(headers=[])
NEED_DATA
Traceback (most recent call last):
  File ".dev/socks_proxy.py", line 44, in <module>
    main()
  File ".dev/socks_proxy.py", line 40, in main
    asyncio.run(socks_proxy(proxy_server, proxy_mode))
  File "/Users/yeray/.pyenv/versions/3.8.1/lib/python3.8/asyncio/runners.py", line 43, in run
    return loop.run_until_complete(main)
  File "/Users/yeray/.pyenv/versions/3.8.1/lib/python3.8/asyncio/base_events.py", line 612, in run_until_complete
    return future.result()
  File ".dev/socks_proxy.py", line 26, in socks_proxy
    http_version, status_code, reason, headers, stream = await http.request(
  File "/Users/yeray/code/personal/_forks/httpcore/httpcore/_async/http_proxy.py", line 92, in request
    return await self._socks4_request(
  File "/Users/yeray/code/personal/_forks/httpcore/httpcore/_async/http_proxy.py", line 225, in _socks4_request
    if connection is None:
  File "/Users/yeray/code/personal/_forks/httpcore/httpcore/_async/connection.py", line 69, in request
    return await self.connection.request(method, url, headers, stream, timeout)
  File "/Users/yeray/code/personal/_forks/httpcore/httpcore/_async/http11.py", line 58, in request
    ) = await self._receive_response(timeout)
  File "/Users/yeray/code/personal/_forks/httpcore/httpcore/_async/http11.py", line 118, in _receive_response
    event = await self._receive_event(timeout)
  File "/Users/yeray/code/personal/_forks/httpcore/httpcore/_async/http11.py", line 146, in _receive_event
    print(event)
  File "/Users/yeray/.pyenv/versions/3.8.1/lib/python3.8/contextlib.py", line 131, in __exit__
    self.gen.throw(type, value, traceback)
  File "/Users/yeray/code/personal/_forks/httpcore/httpcore/_exceptions.py", line 12, in map_exceptions
    raise to_exc(exc) from None
httpcore._exceptions.ProtocolError: can't handle event type ConnectionClosed when role=SERVER and state=SEND_RESPONSE
</details>

Any ideas?